Learner dashboard 3.2

arunabh98 opened the pull request:
Things left - 
Add the learner dashboard icons to the search page.
Add the end to end test - I am facing some difficulty in this as this feature is very dynamic.

arunabh98 committed:
Init work.

arunabh98 committed:
Learner dashboard milestone 3.1

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
Add tests.

arunabh98 committed:
Partially address review comments.

arunabh98 committed:
Combine handlers.

arunabh98 committed:
Add tests.

arunabh98 committed:
Remove the old handlers.

arunabh98 committed:
Fix function name.

arunabh98 committed:
Address review comments 2.

arunabh98 committed:
Remove redundant code.

arunabh98 committed:
Address  another review comment.

arunabh98 committed:
Fix nerge conflicts.

arunabh98 committed:
Add play later to option.

arunabh98 committed:
Add play later icon.

arunabh98 committed:
Merge branch 'learner-dashboard-sixth-milestone' into learner-dashboard-3.2

arunabh98 committed:
Add learner dashboard icons to library.

arunabh98 committed:
Add remove modal.

arunabh98 committed:
Add playlist to dashboard.

arunabh98 committed:
Fix merge conflicts.

arunabh98 committed:
Add backend feedback.

arunabh98 committed:
Remove unncessary code.

arunabh98 committed:
Fix rights manager.

arunabh98 committed:
Merge remote-tracking branch 'upstream/develop' into learner-dashboard-3.2

arunabh98 committed:
Add list.

arunabh98 committed:
Fix big in sortable list.

arunabh98 committed:
Further progress.

arunabh98 committed:
Fix part of e2e test.

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
Add new line.

arunabh98 committed:
Remove newline.

arunabh98 committed:
Remove unnecessary changes.

arunabh98 committed:
Add new line.

arunabh98 committed:
Fix rights manager.

arunabh98 committed:
Streamline exploration tile.

arunabh98 committed:
Add collections and search to playlist.

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
Fix more lint errors.

arunabh98 committed:
Fix errors.

arunabh98 committed:
Fix lint errors.

jaredsilver reviewed (commented):
Something weird is happening when I drag an exploration to change its order (this is in Chrome):

https://cl.ly/m44x

jaredsilver reviewed (commented):
What is the behavior when I attempt to add an exploration that is in my creator dashboard? I seem to be getting an information toast in the bottom right saying that the exploration is in my creator dashboard, but the icon on the exploration says that it's been added to my playlist (which I don't believe it has).

Edit: it seems like the same behavior occurs when the play later list is full. I see the information toast, but the icon tooltip still shows it as having been added.

Unrelated: is there a way to make that informational toast last longer? It's tough to read the text for the "playlist is full" message because it disappears rather quickly.

jaredsilver reviewed (changes requested):
It appears there is no way to access the play later list if there is not an exploration completed or in progress. It seems important that users be able to access their play later list before having an exploration completed or in progress. Can we allow users past the empty state on the learner dashboard if they have at least one exploration/collection in their play later list?

Alternatively, could we display the play later list in the empty state if there is at least one exploration/collection in it?

jaredsilver reviewed (changes requested):
It seems that the design scales oddly at certain viewport widths:

<img width="290" alt="screen shot 2017-08-17 at 5 38 07 pm" src="https://user-images.githubusercontent.com/1312199/29434921-eb6bd256-8372-11e7-94d2-8873edf60c18.png">

<img width="606" alt="screen shot 2017-08-17 at 5 37 54 pm" src="https://user-images.githubusercontent.com/1312199/29434920-eb694edc-8372-11e7-833d-e8dc22bb6d4f.png">

jaredsilver commented:
Only other thing is that the numbers (and, on hover, handlebars) appear to be missing from the mocks. Aside from that, this looks awesome! Thanks @Arunabh98!

seanlip commented on assets/i18n/en.json:
```diff
@@ -202,16 +202,22 @@
     "I18N_LANGUAGE_FOOTER_VIEW_IN": "View Oppia in:",
     "I18N_LEARNER_DASHBOARD_COMPLETED_SECTION": "Completed",
     "I18N_LEARNER_DASHBOARD_COMPLETED_TO_INCOMPLETE_COLLECTIONS": "<[numberMoved]> of the collections you completed have been moved to the 'in progress' section as new explorations have been added to them!",
+    "I18N_LEARNER_DASHBOARD_DELETED_COLLECTIONS_FROM_PLAYLIST": "<[numberDeleted]> of the collections in your 'PLay Later' list have been recently deleted. We are sorry for the inconvenience",
```
PLay Later is miscapitalized. Ditto below.

Also, I suggest "have been deleted by the owner."

arunabh98 replied:
Done!

seanlip commented on assets/i18n/en.json:
```diff
@@ -202,16 +202,22 @@
     "I18N_LEARNER_DASHBOARD_COMPLETED_SECTION": "Completed",
     "I18N_LEARNER_DASHBOARD_COMPLETED_TO_INCOMPLETE_COLLECTIONS": "<[numberMoved]> of the collections you completed have been moved to the 'in progress' section as new explorations have been added to them!",
+    "I18N_LEARNER_DASHBOARD_DELETED_COLLECTIONS_FROM_PLAYLIST": "<[numberDeleted]> of the collections in your 'PLay Later' list have been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_COLLECTION_FROM_PLAYLIST": "<[numberDeleted]> of the collections in your 'PLay Later' list has been recently deleted. We are sorry for the inconvenience",
```
I think we have existing examples in i18n of using a pluralized form. Let's follow those, instead of making multiple separate keys.

arunabh98 replied:
Done!

seanlip commented on assets/i18n/en.json:
```diff
@@ -237,6 +245,9 @@
     "I18N_LEARNER_DASHBOARD_SUGGESTION_TEXT": "Suggestion",
     "I18N_LEARNER_DASHBOARD_VIEW_SUGGESTION": "View Suggestion",
+    "I18N_LIBRARY_ACTIVITY_COMPLETED_ICON": "You have completed this",
+    "I18N_LIBRARY_ACTIVITY_IN_LEARNER_PLAYLIST": "Already added to Play later",
```
How about:
- Already added to playlist
- Add to Playlist

arunabh98 replied:
Done!

seanlip commented on assets/i18n/en.json:
```diff
@@ -287,6 +298,7 @@
     "I18N_LIBRARY_GROUPS_SCIENCE" : "Science",
     "I18N_LIBRARY_GROUPS_SOCIAL_SCIENCE": "Social Science",
     "I18N_LIBRARY_GROUPS_TOP_RATED_EXPLORATIONS": "Top-Rated Explorations",
+    "I18N_LIBRARY_INCOMPLETE_ACTIVITY_ICON": "You are completing this",
```
Hm. Perhaps: "You've started this exploration, but haven't finished it yet." or "You're partway through this exploration." ("In progress" != "(almost) completing".)

arunabh98 replied:
Done!

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -202,16 +202,22 @@
 	"I18N_LANGUAGE_FOOTER_VIEW_IN": "Text for the site language selector. - When the user selects a different language, the site (not the explorations) is translated into that language.",
 	"I18N_LEARNER_DASHBOARD_COMPLETED_SECTION": "Text for the completed section in the learner dashboard.",
 	"I18N_LEARNER_DASHBOARD_COMPLETED_TO_INCOMPLETE_COLLECTIONS": "Text displayed in the learner dashboard. - The text is shown when new exporations are added to the collections the user has already completed.",
+	"I18N_LEARNER_DASHBOARD_DELETED_COLLECTIONS_FROM_PLAYLIST": "Text displayed in the learner dashboard. - This text is shown when more than one collections in the playlist have been deleted.",
```
collections --> collection

arunabh98 replied:
Done!

seanlip commented on core/controllers/learner_playlist.py:
```diff
@@ -27,12 +27,31 @@ class LearnerPlaylistHandler(base.BaseHandler):
         position_to_be_inserted_in = self.payload.get('index')
+
+        playlist_limit_exceeded = False
+        belongs_to_completed_or_incomplete_list = False
```
Nit: declare this in the same order as in the tuple below.

seanlip replied:
That said, on reading further, I'm not sure this is the right modelling. It looks like you should be returning some kind of error object/constant according to the error you actually want to show in the dashboard. This would also be cleaner than returning tuples all over the place.

arunabh98 replied:
I am not exactly thinking of this as errors but more of as informational texts to the user. Also, that being said all three of these reflect different aspects are not really related to the each other so I'm not sure if it's a good idea to group them together - For example these three bools are calculated at different points in the code again grouping them into an object is not really a good idea. What do you think?

seanlip replied:
Let's say we always prioritize error message (or informational text) A over error message B. If A occurs, does it make any difference to the frontend whether B occurs?

If not, then really the piece of info to pass to the frontend is just the single prioritized message.

arunabh98 replied:
No we don't prirotitize the messages. I have changed the code accordingly. All the information text are important. If the user fails to add an exploration to the playlist he/she should know all the information. Or else if the user tries to add an exploration that he/she has created, but is not added because 1) it is created by him/her. 2) the playlist limit has exceeded. If we only show the first message the user will try to add another exploration instead of completing activities in the playlist.

seanlip commented on core/domain/learner_playlist_services.py:
```diff
@@ -92,22 +92,32 @@ def mark_exploration_to_be_played_later(
     learner_playlist = get_learner_playlist_from_model(
         learner_playlist_model)
 
+    playlist_limit_exceeded = False
```
Here and below, the Returns docstring needs updating.

arunabh98 replied:
Done!

seanlip commented on core/controllers/learner_playlist.py:
```diff
@@ -27,12 +27,31 @@ class LearnerPlaylistHandler(base.BaseHandler):
     @acl_decorators.can_access_learner_dashboard
```
Here and elsewhere: tests?

arunabh98 replied:
I have added the tests for learner playlist in the last PR right? Are you asking the tests for @acl_decorators.can_access_learner_dashboard? I am assuming that decorator works correctly and just using it as a module.

seanlip replied:
I'm talking about tests for the new code that's been added to post().

arunabh98 replied:
Done!

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -324,37 +360,32 @@ def add_exp_to_learner_playlist(
             end.
+
+    Returns:
+        bool. It is true if the exploration being added belongs to the completed
```
This docstring seems wrong, and I suggest you return some kind of error object/constant instead.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/playlist_tile/CollectionPlaylistTileDirective.js:
```diff
@@ -0,0 +1,76 @@
+        'collection_playlist_tile_directive.html'),
+      controller: [
+        '$scope', 'oppiaDatetimeFormatter',
+        'COLLECTION_VIEWER_URL', 'COLLECTION_EDITOR_URL', function($scope,
```
Break after '(' and have the line breaks in the deps be at the same locations as those in the stringified list.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/playlist_tile/CollectionPlaylistTileDirective.js:
```diff
@@ -0,0 +1,76 @@
+            );
+          };
+
+          $scope.getCompleteThumbnailIconUrl = function () {
```
Here and below, drop the space before '('.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/playlist_tile/collection_playlist_tile_directive.html:
```diff
@@ -0,0 +1,41 @@
+<md-card class="oppia-activity-summary-tile oppia-activity-playlist-tile">
```
Sorry, no, this won't do. 90+% of the code here duplicates that of the summary tile.

If this is meant to be a replica of the summary tile, then consider reusing the existing tile (customizing it slightly if needed). If this is meant to be a completely different thing altogether, then superfluous functionality should be removed entirely and the data needed stripped down to just the bare minimum. But we cannot have something that's sort of a hybrid of the summary tile and the playlist tile. That's just conceptually muddy.

Ditto for the exploration version.

arunabh98 replied:
I have deleted these file! :D Now I am reusing the existing tile and customizing only slightly.

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -371,25 +424,53 @@ oppia.controller('LearnerDashboard', [
-                    activityId: activity.id
-                  }));
-                $http['delete'](removeCollectionUrl);
+              if (sectionNameI18nId ===
```
I have a feeling you can find a cleaner way to do this that involves much less duplication and cyclomatic complexity. For example, activityType just depends on the subsectionName., so just do a quick calculation at the outset to see what it should be. I don't think you need a lot of branching here and it should be possible to do this with just a single call to UrlInterpolationService and a single call to $http['delete'] in the whole thing.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -400,21 +481,39 @@ oppia.controller('LearnerDashboard', [
-          var index = $scope.incompleteExplorationsList.indexOf(activity);
-          if (index !== -1) {
-            $scope.incompleteExplorationsList.splice(index, 1);
+        if (sectionNameI18nId ===
```
Same thing. This is too complicated.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -20,14 +20,20 @@ oppia.directive('activityTilesInfinityGrid', [
     return {
       restrict: 'E',
+      scope: {
+        learnerDashboardActivityIds: '=learnerDashboardActivityIds',
```
Use & if you don't need to modify the parent var.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -20,14 +20,20 @@ oppia.directive('activityTilesInfinityGrid', [
-        function($scope, $rootScope, searchService) {
+        '$scope', '$rootScope', '$http', 'searchService', 'alertsService',
+        '$modal', function(
+          $scope, $rootScope, $http, searchService, alertsService, $modal) {
```
Line breaks should follow stringified list.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -20,14 +20,20 @@ oppia.directive('activityTilesInfinityGrid', [
+          $scope, $rootScope, $http, searchService, alertsService, $modal) {
           $scope.endOfPageIsReached = false;
           $scope.allActivitiesInOrder = [];
+          var hoverOverActivity = false;
```
Seems like a verb. I'm not sure what this represents. Is it whether an activity is hovered over? If so, why not remove this and just use activeActivityId (or, perhaps, hoveredActivityId) to track this?

arunabh98 replied:
Changed the variable names to more appropriate ones. Now the names more clearly explain the logic.

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -38,6 +44,137 @@ oppia.directive('activityTilesInfinityGrid', [
+                collectionPlaylistIds.indexOf(activityId) !== -1) {
+              return false;
+            } else {
+              return hoverOverActivity && (activeActivityId == activityId);
```
===

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -38,6 +44,137 @@ oppia.directive('activityTilesInfinityGrid', [
+                if (response.data.belongs_to_completed_or_incomplete_list) {
+                  alertsService.addInfoMessage(
+                    'You have already completed or are completing this ' +
+                    ' activity.');
```
Extra space at start of string.

arunabh98 replied:
Fixed!

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -38,6 +44,137 @@ oppia.directive('activityTilesInfinityGrid', [
+                    'and remove some.');
+                } else {
+                  alertsService.addSuccessMessage(
+                    'Succesfully added to your \'Play Later\' list.');
```
"Successfully" is misspelled.

arunabh98 replied:
Fixed!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -27,13 +27,14 @@ oppia.constant('LIBRARY_PAGE_MODES', {
+  'i18nIdService', 'urlService', 'ALL_CATEGORIES', 'searchService',
+  'windowDimensionsService', 'UrlInterpolationService', 'LIBRARY_PAGE_MODES',
+  'LIBRARY_TILE_WIDTH_PX', 'alertsService', function(
+      $scope, $http, $modal, $rootScope, $window, $timeout, i18nIdService,
```
Ditto re line breaks.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -113,6 +118,131 @@ oppia.controller('Library', [
+          var index = (
+            $scope.learnerDashboardActivityIds.exploration_playlist_ids.indexOf(
+              activityId));
+          if (index !== -1) {
```
Could the "=== -1" case happen in practice? If not, should an error be thrown in that circumstance?

arunabh98 replied:
This cannot happen in practice. Just introduced this as an extra check. If this happens it means that the user has clicked on the 'x' button on an activity that does not exist.

seanlip commented on core/templates/dev/head/pages/library/SearchResultsDirective.js:
```diff
@@ -20,6 +20,9 @@ oppia.directive('searchResults', [
     return {
       restrict: 'E',
+      scope: {
+        learnerDashboardActivityIds: '=learnerDashboardActivityIds',
```
Actually, passing this info through all the directives like this is rather ugly.

Here's what I suggest doing. The Library controller should, when the data is retrieved from the backend, initialize a service with this data. The job of that service is to keep track of the learner-dashboard-related stuff, and do nothing else. The service should have methods that allow modification of its internal state. There should be a backendApi service in templates/dev/head/domain that handles updates to the backend.

If sub-directives or components need to refer to the learner dashboard info service, they can do so directly. This way, you can also test that service using Karma tests.

Conceptually, it looks like all the add/remove from learner playlist stuff is a function of just the summary tile. The library controller and the search controller should not need to know about it.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/activity_tiles_infinity_grid_directive.html:
```diff
@@ -1,7 +1,33 @@
      infinite-scroll-disabled="endOfPageIsReached || searchResultsAreLoading">
-  <div ng-repeat="activity in allActivitiesInOrder track by $index" style="display: inline-block;">
+  <div ng-repeat="activity in allActivitiesInOrder track by $index" style="display: inline-block;" ng-mouseenter="setHoverOverActivity(activity.id)" ng-mouseleave="setHoverOverActivity(activity.id)">
+    <i class="protractor-test-add-exp-to-playlist-icon oppia-learner-dashboard-icon fa fa-clock-o"
```
Shouldn't all these icons just be part of the summary tile? You could, on initialization of the summary tile directive, check if the learner dashboard info service exists, and, if so, show these icons. Or the parent directive can pass in a boolean dictating whether to show these extra things or not.

For example, it doesn't seem unreasonable to add this functionality to summary tiles in the profile page.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -204,6 +230,23 @@ <h2 ng-class="{'active': activeGroupIndex === $index}" class="oppia-library-grou
     </div>
   </div>
 
+  <script type="text/ng-template" id="modals/removeActivity">
```
Pull this to a separate file so it can be retrieved asynchronously as needed. See example [here](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/exploration_editor/ExplorationEditor.js#L373) and [here](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/exploration_editor/welcome_modal_directive.html) for the new way of doing this.

arunabh98 replied:
Done!

seanlip reviewed (changes requested):
Thanks, @Arunabh98 -- took a pass. I feel like there's a lot of copy/paste going on in this PR, it would be good to find ways to simplify. Otherwise, there's too much code to maintain, and divergences/discrepancies are bound to occur, going forward.

Feel free to set up some time with me if you want to talk through any design decisions.

rachelwchen commented:
When I added an exploration to "Play Later," the pop-up text says "this is present in your creator dashboard" instead of "this is present in your learner dashboard."
![111](https://user-images.githubusercontent.com/12159451/29477843-2fecd572-8438-11e7-807c-6224669a5cff.png)

rachelwchen commented:
On mobile, the alignment of exploration title seems off -- there's more padding on the left than on the right.
![112](https://user-images.githubusercontent.com/12159451/29478547-5fab2946-843b-11e7-8b3f-8eb166478d9a.png)

rachelwchen commented:
Hm, what is the expected behavior in saving exploration progress?
When I landed on card 3, left the exploration, and accessed it again through my learner dashboard -- it took me back to card 1, with no answers saved. Is this expected to happen in dev mode?
<img width="1246" alt="screen shot 2017-08-18 at 5 35 36 pm" src="https://user-images.githubusercontent.com/12159451/29478647-e81a656c-843b-11e7-9add-2e349d0cfa02.png">

-- Left the exploration, and came back to it through the learner dashboard --
<img width="1229" alt="screen shot 2017-08-18 at 5 35 52 pm" src="https://user-images.githubusercontent.com/12159451/29478654-f4d61454-843b-11e7-8341-93260cccbb3d.png">

arunabh98 committed:
Address review comments.

arunabh98 committed:
Address review comments.

arunabh98 committed:
Same for collections.

arunabh98 committed:
Address more review comments.

arunabh98 commented:
I have fixed the weird dragging behavior and although it does not happen most of the times, it may happen rarely. But that is also easily fixed by reloading the page. The toast behavior has been fixed too. 

Users can now access the playlist on the learner dashboard if they have at least one exploration/collection in it.

The design scaling problems have also been fixed for almost all viewports.

Also @jared, the handles are now irrelevant as the user can now click anywhere on the tile and drag it. So there is no point of adding a handle as it would not really be a handle for dragging the activity. Also, numbers are not that important as there are only 10 activities in the dashboard.

arunabh98 commented:
I have added a service to fetch the learner dashboard ids instead of passing ids around which I agree - was ugly. Now to fetch the ids we just have to use that service. The learner playlist functionality has been shifted to the activity directive as conceptually it should be part of the activity itself. The functions in the collection directive and exploration directive are not the same and they have customized for collections/explorations thus ensuring no code duplication is happening.

arunabh98 commented:
I believe this is ready for another review.

arunabh98 committed:
Address all review comments.

arunabh98 committed:
Fix lint errors.

seanlip commented:
Please reply to individual comments. The first suggestion I had when viewing the changes was to use the standard pluralized form for the i18n files, but then I noticed I made that comment in the last round and it has not been addressed. There is info about this on the wiki.

jaredsilver reviewed (changes requested):
Aside from some styling stuff of the variety Rachel mentioned (alignment and spacing being off in some places at certain viewports), I only see two other issues:

1. There is a weird behavior with the dragging in Chrome where it will always automatically scroll down to the bottom of the page when you're clicking and holding an exploration. To reproduce, add 10 explorations to the playlist, and then click and drag the first exploration down a space. It will continue scrolling, and you won't be able to place it where you want to place it.

2. At a certain viewport width (between when the card text is hidden and when it appears again because the menu jumps up top), the last card's text still appears like so:

<img width="919" alt="screen shot 2017-08-23 at 8 22 18 pm" src="https://user-images.githubusercontent.com/1312199/29644036-37ab859e-8841-11e7-8987-ddd4c1a235dd.png">

Further, when you rearrange them, the text is visible underneath.

arunabh98 commented:
Oh, I must have missed that by mistake, @seanlip. Anyway, I have addressed it now.

@jaredsilver I'll address those in a few hours. 

Till now, it is ready for another code review.

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -202,16 +202,22 @@
 	"I18N_LANGUAGE_FOOTER_VIEW_IN": "Text for the site language selector. - When the user selects a different language, the site (not the explorations) is translated into that language.",
 	"I18N_LEARNER_DASHBOARD_COMPLETED_SECTION": "Text for the completed section in the learner dashboard.",
 	"I18N_LEARNER_DASHBOARD_COMPLETED_TO_INCOMPLETE_COLLECTIONS": "Text displayed in the learner dashboard. - The text is shown when new exporations are added to the collections the user has already completed.",
+	"I18N_LEARNER_DASHBOARD_DELETED_COLLECTIONS_FROM_PLAYLIST": "Text displayed in the learner dashboard. - This text is shown when more than one collection in the playlist have been deleted.",
```
It seems strange that the changes here are not in sync with those in en.json.

arunabh98 replied:
Done! Although I would have liked this to be pointed out earlier. I remember specifically asking that what should I do in for the plural cases and I was suggested to create different keys. Now I had to go through all the json files (after translation had happened) to bring them in sync with en.json and ensure that they had no extra keys.

seanlip replied:
Ah, thanks for mentioning that. You're right, I had missed that this was an issue when reviewing #3482, and I agree that catching it earlier would have helped. Sorry about that!

seanlip commented on assets/i18n/en.json:
```diff
@@ -202,16 +202,16 @@
-    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_COLLECTIONS": "<[numberDeleted]> of the collections in progress have been recently deleted. We are sorry for the inconvenience",
-    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_EXPLORATION": "<[numberDeleted]> of the explorations in progress has been recently deleted. We are sorry for the inconvenience",
-    "I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_EXPLORATIONS": "<[numberDeleted]> of the explorations in progress have been recently deleted. We are sorry for the inconvenience",
+    "I18N_LEARNER_DASHBOARD_DELETED_COLLECTIONS_FROM_PLAYLIST": "{numberDeleted, plural, one{1 of the collections in your 'Play Later' list has been deleted by the owner. We are sorry for the inconvenience} other{# of the collections in your 'Play Later' list have been deleted by the owner. We are sorry for the inconvenience}}",
```
Please consider doing these in hi.json, too, to make sure they're functioning correctly.

arunabh98 replied:
Done!

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -117,6 +117,38 @@ def get(self):
+        # collections completed by the user, the activities currently being
+        # completed and the ones present in the playlist of the user.
+        activity_ids = (
+            learner_progress_services.get_ids_of_activities_in_learner_dashboard( # pylint: disable=line-too-long
```
Let's call this domain object something like "learner_dashboard_activities" and say learner_dashboard_activities = learner_progress_services.get_learner_dashboard_activities() or similar. "activity_ids" seems like a list of activity IDs and I believe we use it in other contexts to mean that, so the current formulation is slightly confusing.

arunabh98 replied:
Done!

seanlip reviewed (commented):
OK -- I started reviewing this, and came across stuff I thought I had commented on before, and on looking back at the comment history I noticed I had indeed commented on it before and the thread was not replied to. So, in order to avoid unnecessary back-and-forth, I am going to hold off on a further review until you have addressed _all_ existing comments.

Please read through the comment history and ensure that every single outstanding comment is addressed before asking for another review.

arunabh98 committed:
Add e2e test for plurals.

arunabh98 commented:
Added the test cases for the new code I added in learner_playlist.py

arunabh98 commented:
I checked all the review comments and as far as I can see I have addressed all of them. This is ready for another review.

seanlip commented on core/controllers/learner_dashboard.py:
```diff
@@ -117,6 +117,39 @@ def get(self):
+        }
+
+        self.values.update({
+            'learner_dashboard_activity_ids': (
```
Let's do, instead:

    'learner_dashboard_activity_ids': learner_dashboard_activities.to_dict(),

and remove all the stuff above.

arunabh98 replied:
Done!

seanlip commented on core/domain/learner_playlist_services.py:
```diff
@@ -79,6 +79,12 @@ def mark_exploration_to_be_played_later(
             end.
+
+    Returns
+        bool. Indicates whether the playlist limit of the user has been
```
"Returns" should only have one thing. Say:

```
   Returns:
        (bool, bool). The first boolean indicates ..., and the second boolean indicates ...
```

Ditto below.

arunabh98 replied:
Done!

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -311,6 +311,50 @@ def mark_collection_as_incomplete(user_id, collection_id):
+            end.
+
+    Returns:
+        bool. Indicates whether the collection belongs to the completed or
```
Seems unclear, from reading this, which state corresponds to True and which to False. Maybe just have one condition "whether X" (as opposed to "whether X or Y").

EDIT: ah, I misread, I think. Maybe write: "... whether the collection already exists in either of the "completed collections" or "incomplete collections" lists".

arunabh98 replied:
Done!

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -311,6 +311,50 @@ def mark_collection_as_incomplete(user_id, collection_id):
+        bool. Indicates whether the collection belongs to the created or edited
+            collections of the user.
+    """
+    completed_collection_ids = get_all_completed_collection_ids(user_id)
```
There are two calls here resulting in two datastore RPCs. Can we just have one?

arunabh98 replied:
I can but we already have functions for fetching completed and incomplete collections and I don't think this would create a big difference in the user experience. It's not like the learner dashboard which is extensive and the user actually has to wait for the page to load. In this case, this is just a user action for adding something. So I believe we would be sacrificing simplicity and easy to understand code for just for fetching multiple models at once.

 What do you think?

seanlip replied:
Well, the right fix is to find a way to do it in such a way that keeps the code simple and easy-to-understand, _and_ fetch multiple models at once :) For example, finding a way to conceptualize "incomplete collection IDs" and "completed collection IDs" in a single object.

But I agree with your reasoning that there's a tradeoff here, and that this isn't a huge problem in the larger scale of things. So, for this PR, let's consider this change optional, but it would be nice to find a way to streamline it a bit more in the future.

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -850,29 +893,17 @@ def get_collection_summary_dicts(collection_summaries):
-        list(str). The titles of the collections to which new explorations have
-            been added.
-
+        activity_ids. The domain object containing the ids of all activities
```
The first thing here should be the type of the return value, not its name.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -34,17 +34,22 @@ oppia.directive('collectionSummaryTile', [
         isLinkedToEditorPage: '=?isLinkedToEditorPage',
-        getCategory: '&category'
+        getCategory: '&category',
+        playlistMode: '=?playlistMode',
```
I'm not sure you need the two-way binding here. Could you use &? Ditto below.

(I think isLinkedToEditorPage shouldn't be two-way-bound here either, but it's up to you whether you want to change that or not.)

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -70,6 +75,135 @@ oppia.directive('collectionSummaryTile', [
+          var activeActivityId = '';
+
+          $scope.setCurrentlyHoveringOverActivity = function(activityId) {
+            activeActivityId = activityId;
```
This whole tile represents just one activityId. So shouldn't we be keeping track of just activityIsActive perhaps? Or do we even need a separate variable at all, since then currentlyHoveringOverActivity would also track the same thing?

arunabh98 replied:
Now just tracking if activity is active.

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -70,6 +75,135 @@ oppia.directive('collectionSummaryTile', [
+          $scope.showAddToLearnerPlaylistIcon = function(activityId) {
+            if ($scope.learnerDashboardActivityIds) {
+              var incompleteCollectionIds = (
+                $scope.learnerDashboardActivityIds.incomplete_collection_ids);
```
Shouldn't learnerDashboardActivityIds be a frontend domain object? We shouldn't be referring to backend dicts with snake_case throughout the frontend.

You can then put some of this logic into the domain object itself, e.g. $scope.learnerDashboardActivityIds.containsCollection(collectionId).

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -34,17 +34,22 @@ oppia.directive('collectionSummaryTile', [
+          COLLECTION_VIEWER_URL, COLLECTION_EDITOR_URL,
+          UrlInterpolationService, alertsService) {
           $scope.DEFAULT_EMPTY_TITLE = 'Untitled';
+          $scope.collectionType = constants.ACTIVITY_TYPE_COLLECTION;
```
I don't see why you need to define this here and pass it around. Shouldn't the functions within this directive's controller already be aware that they're operating within the context of a collection?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -70,6 +75,135 @@ oppia.directive('collectionSummaryTile', [
+            currentlyHoveringOverActivity = !currentlyHoveringOverActivity;
+          };
+
+          $scope.showAddToLearnerPlaylistIcon = function(activityId) {
```
This returns a boolean, so the method name is misleading; it sounds like it does something. Consider something like $scope.canCollectionBeAddedToLearnerPlaylist().

Also, throughout this directive controller, use collectionId instead of activityId (and similarly for other uses of "activity"). This is a _collection_ summary tile and not a generic activity summary tile.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -70,6 +75,135 @@ oppia.directive('collectionSummaryTile', [
+            currentlyHoveringOverActivity = !currentlyHoveringOverActivity;
+          };
+
+          $scope.showAddToLearnerPlaylistIcon = function(activityId) {
```
There's still way too much duplication in this and the following two methods (with ExplorationSummaryTile). I recommend the following:

- create a learnerDashboardActivityIds _domain object_ and use that domain object to make the code in showAddToLearnerPlaylistIcon() minimal (see below)
- create a LearnerPlaylistService with "add"/"remove"/"does X exist" methods; this service calls the BackendApiService to do a fetch as soon as it gets initialized. The service also depends on the corresponding BackendApiService and sends HTTP requests to it for "add"/"remove", and updates its stored domain object accordingly (a bit like the ExplorationData service in the exploration editor). The service should throw an error if it's asked to add/remove before it has been initialized, and return false if a lookup is made before it has been initialized. You can tie methods in this directive to that service using something like `$scope.canCollectionBeAddedToLearnerPlaylist = LeranerPlaylistService.canCollectionBeAdded`, so that the return values update automatically as the data changes.
- include this LearnerPlaylistService as a dependency here, and remove the corresponding arg that's being passed in from the parent directive.

That way, you can use this service in both directives, with minimal code in the latter. The service can also be easily tested.

arunabh98 replied:
Created a service!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -70,6 +75,135 @@ oppia.directive('collectionSummaryTile', [
+                }
+              });
+
+            if (activityType == constants.ACTIVITY_TYPE_COLLECTION) {
```
===

arunabh98 replied:
Fixed since its no longer present.

seanlip commented on core/templates/dev/head/components/summary_tile/collection_summary_tile_directive.html:
```diff
@@ -1,12 +1,40 @@
-<md-card class="oppia-activity-summary-tile">
+<span ng-if="learnerDashboardActivityIds">
+<i class="protractor-test-add-exp-to-playlist-icon oppia-learner-dashboard-icon fa fa-clock-o"
```
This and the following lines should be indented by 2 from the previous line. Ditto for the exploration summary tile.

In fact, I suggest you pull the next 26 lines or so into a separate directive that's initialized with the activity type and activity ID, since it's again duplicated between both the summary tiles. That directive could then depend on LearnerPlaylistService and handle all the button clicks directly without bothering the collectionSummaryTile/explorationSummaryTile.

arunabh98 replied:
I tried this, and I don't think this would be a good idea as this piece is dynamic and dependent on the parent. So I'm facing a lot of trouble in interfacing this. Especially the hovering part.

seanlip replied:
If you had a service that was directly injected into the directive and provided the playlist data, then the directive can take care of its own hovering state, right? Not sure why the parent needs to know...

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -250,6 +251,58 @@ oppia.controller('LearnerDashboard', [
+      'ui-floating': false 
+    };
+
+    $scope.collectionPlaylistSortableOptions = {
```
Again, too much duplication, given that pretty much the only thing that's different is the activity type. Either find a way to pass that cleanly along with the item, or do something like:

```
var getPlaylistSortableOptions(activityType) = ...;

$scope.collectionPlaylistSortableOptions = getPlaylistSortableOptions(constants.ACTIVITY_TYPE_COLLECTION);
$scope.explorationPlaylistSortableOptions = getPlaylistSortableOptions(constants.ACTIVITY_TYPE_EXPLORATION);
```

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -371,26 +426,37 @@ oppia.controller('LearnerDashboard', [
+                  LEARNER_DASHBOARD_SECTION_I18N_IDS.PLAYLIST) {
+                /* eslint-disable max-len */
+                removeActivityUrl = (
                   UrlInterpolationService.interpolateUrl(
```
Remove the duplication. Set a variable with the URL template based on the if/else, and make just a single call to UrlInterpolationService.interpolateUrl().

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -400,21 +466,39 @@ oppia.controller('LearnerDashboard', [
-          if (index !== -1) {
-            $scope.incompleteCollectionsList.splice(index, 1);
+        } else if (sectionNameI18nId ===
+                   LEARNER_DASHBOARD_SECTION_I18N_IDS.PLAYLIST){
```
Add space before {

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -24,11 +24,14 @@ oppia.directive('activityTilesInfinityGrid', [
-        '$scope', '$rootScope', 'searchService',
-        function($scope, $rootScope, searchService) {
+        '$scope', '$rootScope', '$http', 'searchService', 'alertsService',
+        '$modal', 'LearnerDashboardIdsBackendApiService', function(
```
$-dependencies should go at the front.

arunabh98 replied:
Fixed!

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -24,11 +24,14 @@ oppia.directive('activityTilesInfinityGrid', [
           $scope.endOfPageIsReached = false;
           $scope.allActivitiesInOrder = [];
-
+          var currentlyHoveringOverActivity = false;
```
Why are we still maintaining all this stuff here?

arunabh98 replied:
Removed!

seanlip commented on core/templates/dev/head/pages/library/ActivityTilesInfinityGridDirective.js:
```diff
@@ -38,6 +41,18 @@ oppia.directive('activityTilesInfinityGrid', [
+            }
+          );
+
+          $scope.setCurrentlyHoveringOverActivity = function(activityId) {
```
Ditto, why is this here?

arunabh98 replied:
Removed!

seanlip commented on core/templates/dev/head/pages/library/LearnerDashboardIdsBackendApiService.js:
```diff
@@ -0,0 +1,30 @@
+// limitations under the License.
+
+/**
+ * @fileoverview get the activity ids that are present in the learner
```
This is more like a method docstring....

What is this BackendApiService for, conceptually? Does it just serve the learner dashboard or does it serve anything which interacts with the learner playlist? If the former, LearnerDashboardBackendApiService is a good name, and this should handle all the backend API calls from the learner dashboard page. If the latter, then perhaps LearnerPlaylistBackendApiService could work, and this should handle all the add/remove calls from the different pages that might show summary tiles. The fileoverview docstring should make this conceptualization clear.

arunabh98 replied:
This file is neither related to the learner dashboard backend not it is related to the learner playlist. It is just for fetching the ids that are present in the learner dashboard. These ids can then be used to display icons, for playlist purposes etc. I have updated the file overview accordingly.

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -125,7 +126,8 @@ <h2 ng-class="{'active': activeGroupIndex === $index}" class="oppia-library-grou
                                              thumbnail-icon-url="tile.thumbnail_icon_url"
-                                             thumbnail-bg-color="tile.thumbnail_bg_color">
+                                             thumbnail-bg-color="tile.thumbnail_bg_color"
+                                             learner-dashboard-activity-ids="learnerDashboardActivityIds">
```
This isn't really the right place to pass this in (see above). At most, pass in an include-library-playlist-icons boolean saying whether you want the tile to show the icons. Whether the icons actually get displayed, and which icons get displayed, should be a result of the summary tile talking to the LearnerPlaylistService.

arunabh98 replied:
I didn't quite understand this. I have created a learner playlist service but can't we pass the ids from here itself? In whatever page we want to show the icon we just need to fetch the ids using LearnerDashboardIdsBackendService and then pass this data as a field to the summary. So if you think about it, passing this info itself acts like a bool whether to display the icons or not. If we do want to display icons we do not pass this information.

And I think this is easily extendable and maintainable also. If we want to display the icons on some page in the future - just fetch the ids and pass this info to the summary. 

How does this sound?

seanlip replied:
I think the answer is still no. Passing data through a long chain of functions is a bit of a code smell. If you added additional sub-directives (between the activity grid and the library, say) or changed the layout of the page or added some other thing, then all the intermediate layers would have to carry the data along. It's also a bit like a global variable that everyone can edit -- you now have this info stored in the service, and in the top-level controller, and in all the subdirectives, and all of these have to be in sync. A few iterations of this will turn the whole page into a big ball of mud as additional changes get made in the future.

The proposal instead is to do everything using the service as a reference point, so that we have the canonical version of the data stored in exactly one place. This also avoids a lot of passing the data around and also avoids outer controllers having to hold more state than needed. Better for one service to take care of all that (i.e. initial load, and subsequent modifications).

Now, you may be wondering, what's the point of the boolean then, and why is a boolean OK when the list of activity IDs is not so OK? Well, that's because the boolean is essentially a UI thing -- it indicates whether or not to display that info in the tile, regardless of whether we actually have the data or not. A "false" flag means we never display it (maybe this is done because of the particular design of the tile); a "true" flag means we can display it if we have the data. This is a direct result of the context in which the directive is placed, and is something that will vary with context, so it's reasonable to re-specify that flag each time because the context is different each time. But the data will always be the same for that particular user.

arunabh98 replied:
This is done!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -74,6 +77,15 @@ oppia.controller('Library', [
       $http.get('/libraryindexhandler').success(function(data) {
         $scope.libraryGroups = data.activity_summary_dicts_by_category;
 
+        if (data.user_email) {
```
Could you use data.user_id or GLOBALS.userIsLoggedIn? But I don't know why you need all this stuff in the library page anyway. There should probably be no changes to this file in this PR.

arunabh98 replied:
Done. Also, I have provided a justification above.

seanlip commented on core/templates/dev/head/pages/library/activity_tiles_infinity_grid_directive.html:
```diff
@@ -10,7 +10,7 @@
                              objective="activity.objective"
                              thumbnail-icon-url="activity.thumbnail_icon_url"
                              thumbnail-bg-color="activity.thumbnail_bg_color"
-                             style="margin-right: 3px;">
```
Just to check, does this deletion fix a bug? Might be worth doing a git blame to ensure you know why it was added in the first place, to make sure you aren't causing a regression.

arunabh98 replied:
Oh, I might have deleted this by mistake. I certainly didn't intend to do this.

seanlip reviewed (commented):
Hi @Arunabh98, I've done another pass! It looks better before, and the comments on the backend are mostly cosmetic. However, the frontend still has too much duplication of large chunks of code, and I've given some explicit suggestions inline for how to reduce this. PTAL when you get a chance.

arunabh98 committed:
Address review comments 2.

arunabh98 committed:
Add hindi translations.

arunabh98 committed:
Partially address review comments.

arunabh98 committed:
Address more review comments.

arunabh98 committed:
Add domain object.

arunabh98 committed:
Add service.

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
Fix a bug.

arunabh98 committed:
Fix lint errors.

arunabh98 commented:
I have done the necessary changes. Now the Library.js does not have even a single change. The entire functionality is shifted to the playlist service and icons directive and they are being used by the summary tiles for display of icons. 

Also, I've tried to include as many tests as possible but omitted tests for functions that just use standard methods like push, indexOf etc.

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -311,6 +311,50 @@ def mark_collection_as_incomplete(user_id, collection_id):
+            gets inserted at the given position. Otherwise it gets added at the
+            end.
+
+    Returns:
```
Just use one return value. Please check the rest of the PR.

arunabh98 replied:
I have changed this and checked the PR. I found one other occasion where I had to change.

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -34,17 +34,19 @@ oppia.directive('collectionSummaryTile', [
+          oppiaDatetimeFormatter, COLLECTION_VIEWER_URL,
+          COLLECTION_EDITOR_URL) {
           $scope.DEFAULT_EMPTY_TITLE = 'Untitled';
+          $scope.collectionType = constants.ACTIVITY_TYPE_COLLECTION;
```
This is a constant, so call it $scope.ACTIVITY_TYPE_COLLECTION

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/ExplorationSummaryTileDirective.js:
```diff
@@ -82,6 +83,7 @@ oppia.directive('explorationSummaryTile', [
           $scope, $http,
           oppiaDatetimeFormatter, RatingComputationService,
           windowDimensionsService) {
+          $scope.explorationType = constants.ACTIVITY_TYPE_EXPLORATION;
```
Ditto.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerDashboardActivityIdsObjectFactory.js:
```diff
@@ -0,0 +1,86 @@
+
+  LearnerDashboardActivityIds.prototype.belongsToLearnerDashboardActivities = (
+    function(
+    activityId) {
```
You should be able to put this after the line above?

That said, the method name is quite long. Maybe just "includesActivity(activityId)"?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerDashboardIconsDirective.js:
```diff
@@ -0,0 +1,94 @@
+          };
+
+          $scope.canActivityBeAddedToLearnerPlaylist = function(
+            activityId) {
```
Move to end of previous line?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerDashboardIconsDirective.js:
```diff
@@ -0,0 +1,94 @@
+                activityId, activityType));
+            if (isSuccessfullyAdded) {
+              /* eslint-disable max-len */
+              $scope.learnerDashboardActivityIds.addToExplorationLearnerPlaylist(
```
I don't understand why we have "exploration" in the name, when there's a general activity type here...

arunabh98 replied:
Fixed!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerPlaylistService.js:
```diff
@@ -0,0 +1,117 @@
+      });
+    };
+
+    var _fetchLearnerDashboardData = function() {
```
Shouldn't this service be maintaining a private copy of the activity ids object? You should only be doing a single fetch.

Or rather: the public API for this service should be something like fetchLearnerDashboardActivityIds(), and the logic in that should be "if data exists, return promise containing data, else make new fetch from backend" or similar.

arunabh98 replied:
Ohh I think this is included in this service by mistake. This is actually the service for fetching the data of the learner dashboard in which case this is correct. I have removed this from here. Sorry for this!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -424,6 +544,8 @@ <h1 class="stat-value"><[subscriptionsList.length]></h1>
             </div>
           </div>
 
+
```
Drop unnecessary blank lines

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -673,6 +778,8 @@ <h3 translate="I18N_LEARNER_DASHBOARD_SUGGESTION_TEXT"></h3>
   <script src="{{TEMPLATE_DIR_PREFIX}}/components/background/BackgroundBannerDirective.js"></script>
   <script src="{{TEMPLATE_DIR_PREFIX}}/components/summary_tile/CollectionSummaryTileDirective.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/domain/learner_dashboard/LearnerPlaylistService.js"></script>
+  <script src="{{TEMPLATE_DIR_PREFIX}}/pages/library/LearnerDashboardIdsBackendApiService.js"></script>
```
This should not go in pages/library. It should go somewhere sensible in domain/. Stuff in pages/ is for just that specific page.

arunabh98 replied:
Done!

seanlip reviewed (commented):
Sorry, I can only take a quick look now. But the overall "shape" is good! I suggest getting a UI review; this is close to the finish line now, code-wise.

arunabh98 committed:
Add directive for icons.

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
Fix more lint errors.

arunabh98 committed:
Remove unnecessary code.

arunabh98 committed:
Remove unnecessary code 2.

arunabh98 committed:
Address review comments 3.

arunabh98 commented:
Addressed review comments.

@jaredsilver I've fixed the issues in dragging elements in most cases. Though sometimes rarely an error might happen. There is an issue about this flakiness opened in ui sortable and after this is sorted even the rare error would not happen.

arunabh98 committed:
Fix e2e tests.

arunabh98 committed:
Remve unintended changes.

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -137,7 +138,8 @@ <h2 ng-class="{'active': activeGroupIndex === $index}" class="oppia-library-grou
                                               thumbnail-bg-color="tile.thumbnail_bg_color"
-                                              class="protractor-test-exp-summary-tile">
+                                              class="protractor-test-exp-summary-tile"
+                                              show-learner-dashboard-icons="userIsLoggedIn">
```
This still shouldn't be passed through. The individual tile should check whether the user is logged in by using e.g. GLOBALS.userIsLoggedIn, and act accordingly. You shouldn't need to go through the library page for that.

The immediate parent of the tile directive might pass a bool attribute along the lines of "show-learner-dashboard-icons-if-possible" (which, if the bool's value is true, conveys that the icons should be shown iff the learner is logged in).

arunabh98 replied:
Okay, I'll just change this. Should I keep the attribute the name still as `show-learner-dashboard-icons`?

seanlip replied:
I think you want the "-if-possible" part I mentioned, for the given reason. Similar comments apply to all the other places where you do this.

arunabh98 replied:
Okay.. cool!

arunabh98 replied:
I have fixed this. Feels good to see so much modularity! :D

arunabh98 committed:
Address review comments 4.

arunabh98 committed:
Fix lint errors.

arunabh98 committed:
Fix more lint errors.

arunabh98 committed:
Merge remote-tracking branch 'upstream/develop' into learner-dashboard-3.2

arunabh98 committed:
Changed another translation file to account for the new keys.

seanlip commented on assets/i18n/sv.json:
```diff
@@ -174,9 +174,7 @@
 	"I18N_INTERACTIONS_SET_INPUT_SUBMIT": "Skicka",
 	"I18N_LANGUAGE_FOOTER_VIEW_IN": "Visa Oppia på:",
 	"I18N_LEARNER_DASHBOARD_COMPLETED_SECTION": "Slutförd",
-	"I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_COLLECTIONS": "<[numberDeleted]> av de pågående samlingarna har nyligen raderats. Vi beklagar besväret",
```
Why delete this?

arunabh98 replied:
Because it did not have the translation for `I18N_LEARNER_DASHBOARD_DELETED_INCOMPLETE_COLLECTION`. So if I had included my key. The other one case would just be blank and it maybe the case we would miss to translate this in the future.

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -913,6 +944,53 @@ def get_activity_progress(user_id):
+    Args:
+        user_id: str. The id of the learner.
+
+    Returns:
```
I feel like I shouldn't have to comment on this in three separate passes, but there are still three return values here.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -34,17 +34,20 @@ oppia.directive('collectionSummaryTile', [
         isLinkedToEditorPage: '=?isLinkedToEditorPage',
-        getCategory: '&category'
+        getCategory: '&category',
+        isPlaylistMode: '&playlistMode',
```
isInPlaylistMode reads better, I think. Or isPlaylistTile. (But this tile isn't a "mode".)

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -34,17 +34,20 @@ oppia.directive('collectionSummaryTile', [
-        getCategory: '&category'
+        getCategory: '&category',
+        isPlaylistMode: '&playlistMode',
+        showLearnerDashboardIcons: '&showLearnerDashboardIcons'
```
I feel like my comments are getting lost somewhere. I thought we were going to do showLearnerDashboardIconsIfPossible (or perhaps showLearnerDashboardIconsIfLoggedIn)?

arunabh98 replied:
I have changed this. Maybe it some bug with GitHub. Now that I'm seeing this it is showing this as outdated.

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -34,17 +34,20 @@ oppia.directive('collectionSummaryTile', [
         'collection_summary_tile_directive.html'),
       controller: [
         '$scope', 'oppiaDatetimeFormatter',
         'COLLECTION_VIEWER_URL', 'COLLECTION_EDITOR_URL', function($scope,
```
Break line after '('. Consider indenting the stuff in '( ... )' by 2 extra spaces to distinguish it from the stuff that's actually in the inner scope.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/ExplorationSummaryTileDirective.js:
```diff
@@ -43,7 +43,9 @@ oppia.directive('explorationSummaryTile', [
         // desktop version of the summary tile is always displayed.
-        mobileCutoffPx: '@mobileCutoffPx'
+        mobileCutoffPx: '@mobileCutoffPx',
+        isPlaylistMode: '&playlistMode',
```
Ditto here and next line re naming of attrs.

arunabh98 replied:
Done! And I had changed this and pushed it. Don't know why that happened.

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -70,6 +73,11 @@ oppia.directive('collectionSummaryTile', [
             return UrlInterpolationService.getStaticImageUrl(url);
           };
+
+          $scope.collectionIsActive = false;
```
Consider $scope.collectionIsCurrentlyHoveredOver -- that might be more clear. "Active" seems like it needs definition.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/ExplorationSummaryTileDirective.js:
```diff
@@ -112,6 +116,11 @@ oppia.directive('explorationSummaryTile', [
 
           $scope.MAX_AVATARS_TO_DISPLAY = 5;
 
+          $scope.explorationIsActive = false;
```
Ditto (naming).

seanlip commented on core/templates/dev/head/components/summary_tile/collection_summary_tile_directive.html:
```diff
@@ -1,12 +1,21 @@
+</learner-dashboard-icons>
+<md-card class="oppia-activity-summary-tile"
+         ng-class="{'oppia-activity-playlist-tile': isPlaylistMode()}"
+         ng-mouseenter="toggleCollectionIsActive()" ng-mouseleave="toggleCollectionIsActive()"
```
Here and in the exp summary tile: instead of toggling, consider having setHoverState(true) and setHoverState(false). That's more explicit/definite, and avoids the assumption of no race conditions between mouseenter and mouseleave.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/exploration_summary_tile_directive.html:
```diff
@@ -1,8 +1,16 @@
+                         activity-active="explorationIsActive">
+</learner-dashboard-icons>
+<md-card ng-class="[isCollectionPreviewTile ? 'oppia-activity-summary-tile-mobile' : 'oppia-activity-summary-tile', {'small-width': !isWindowLarge}, {'oppia-activity-playlist-tile': isPlaylistMode()}]"
+         ng-mouseenter="toggleExplorationIsActive()" ng-mouseleave="toggleExplorationIsActive()"
```
Same thing here re toggle.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -3995,6 +4020,62 @@ md-card.preview-conversation-skin-supplemental-card {
+  text-align: center;
+}
+
+@media(max-width: 895px) {
```
Should these and some of the stuff below also have min-widths in order not to conflict with the rules that apply for smaller widths?

arunabh98 replied:
I have tried and tested this. On reducing the width, the latest one always takes precedence.

seanlip replied:
Yep, I know, but the existing CSS is fragile. Things get reordered. You might want to consider something like

`@media (min-width: 700px) and (max-width: 768px) {`

to make it more robust.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerDashboardActivityIdsObjectFactory.js:
```diff
@@ -0,0 +1,85 @@
+          this.explorationPlaylistIds.indexOf(activityId) !== -1) {
+        return true;
+      } else {
+        false;
```
Careless; no return value. Why is this branch of the code not tested?

arunabh98 replied:
Done! Added tests.

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerDashboardIconsDirective.js:
```diff
@@ -0,0 +1,98 @@
+                 LearnerPlaylistService) {
+          $scope.activityIsActive = true;
+
+          $scope.$watch('activityActive', function(value) {
```
Why do you need to $watch -- doesn't it update automatically? Couldn't you just bind $scope.activityIsActive directly to the attr passed in from the parent?

Another approach might be: the parent passes in an attr along the lines of "parent-tile-is-being-hovered-over", and you can use that as an "&" attr. Then you can maintain a local $scope.iconsAreBeingHoveredOver and also compute $scope.activityIsActive based on whether at least one of the previous two things is true. I think it depends on what exactly you want $scope.activityIsActive to represent (and I suggest choosing a better name -- see above -- and possibly also adding a docstring/comment), but this could use some more clarity.

arunabh98 replied:
Everywhere I've read about this it is advised to use the watch functionality to get this behavior. For example - [https://stackoverflow.com/a/20068832/6331311](https://stackoverflow.com/a/20068832/6331311). Also, I've changed the variable names so hopefully this makes more sense.

seanlip replied:
Hm, OK. I would have thought you could pass them in directly, but I'm not _too_ worried. The concern with $watch is performance -- so maybe try adding a console.log() inside it and see how often it gets called. If it doesn't get called too much then what you have is fine.

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerPlaylistService.js:
```diff
@@ -0,0 +1,112 @@
+      return successfullyAdded;
+    };
+
+    var _removeFromLearnerPlaylist = function(activityId, activityTitle,
```
Here and elsewhere, when stuff in (...) spans more than one line, break after first paren.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/learner_dashboard_icons_directive.html:
```diff
@@ -0,0 +1,23 @@
+<i class="oppia-learner-dashboard-icon fa fa-clock-o"
+   ng-show="canActivityBeAddedToLearnerPlaylist(getActivityId())"
+   ng-click="addToLearnerPlaylist(getActivityId(), getActivityType())"
+   ng-mouseenter="toggleActivityIsActive()" ng-mouseleave="toggleActivityIsActive()"
```
See previous comments re using something more definite than "toggle".

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/learner_dashboard_icons_directive.html:
```diff
@@ -0,0 +1,23 @@
+   tooltip="<['I18N_LIBRARY_ADD_TO_LEARNER_PLAYLIST' | translate]>"
+   tooltip-placement="left"></i>
+<i class="oppia-learner-dashboard-icon fa fa-clock-o"
+   ng-if="learnerDashboardActivityIds.explorationPlaylistIds.indexOf(getActivityId()) !== -1 || learnerDashboardActivityIds.collectionPlaylistIds.indexOf(getActivityId()) !== -1"
```
I just realized that this doesn't make sense to me. Don't you already know the activity type? In that case you should be doing just the appropriate check. Explorations and collections could (unlikely, but it's possible) have the same ID.

I suggest you pull this check out into a standalone function within the directive controller and do it properly based on the activity type. Ditto below.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -250,6 +251,50 @@ oppia.controller('LearnerDashboard', [
+      return {
+        'ui-floating': 'auto',
+        start: function(e, ui) {
+          if(startSort){
```
Add spaces before '(' and after ')'.

arunabh98 replied:
Removed!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -250,6 +251,50 @@ oppia.controller('LearnerDashboard', [
 
+    var getPlaylistSortableOptions = function(activityType) {
+      var windowScrollTop = '';
+      var startSort = false;
```
No idea what this means, and it sounds like a verb. Maybe add comment to explain?

arunabh98 replied:
Done! Found a better solution that prevents even the rare errors I had mentioned earlier.

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -250,6 +251,50 @@ oppia.controller('LearnerDashboard', [
     };
 
+    var getPlaylistSortableOptions = function(activityType) {
+      var windowScrollTop = '';
```
No idea what this means, and slightly surprised it's a string. Maybe add comment to explain?

arunabh98 replied:
Done! Found a better solution that prevents even the rare errors I had mentioned earlier.

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -371,26 +418,32 @@ oppia.controller('LearnerDashboard', [
-                  '/<activityId>', {
-                    activityType: constants.ACTIVITY_TYPE_COLLECTION,
+                activityType = constants.ACTIVITY_TYPE_COLLECTION;
+              }
```
else throw Error

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -371,26 +418,32 @@ oppia.controller('LearnerDashboard', [
+              } else if (sectionNameI18nId ===
+                         LEARNER_DASHBOARD_SECTION_I18N_IDS.INCOMPLETE) {
+                removeActivityUrlPrefix = '/learnerincompleteactivityhandler/';
+              }
```
else throw Error

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -308,15 +320,70 @@ <h1 class="stat-value"><[subscriptionsList.length]></h1>
+                   tooltip="Remove"
+                   tooltip-placement="top"></i>
+                <exploration-summary-tile
+                                        style="height: 145px; display: block;"
```
Move this to end of previous line, and revise the indentation accordingly. It's OK if this causes subsequent lines to go > 80 chars.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/learner_dashboard.html:
```diff
@@ -363,6 +430,62 @@ <h1 class="stat-value"><[subscriptionsList.length]></h1>
+                   tooltip="Remove"
+                   tooltip-placement="top"></i>
+                <collection-summary-tile
+                                         style="height: 145px; display: block;"
```
Ditto.

arunabh98 replied:
Done!

seanlip reviewed (changes requested):
Hi @Arunabh98, I've taken what should be a final thorough pass. Only small comments, but please address them carefully; some comments repeat comments in previous reviews. After this I'm happy to give LGTM from the technical perspective. Thanks!

arunabh98 commented:
Addressed all review comments.

seanlip commented on core/domain/learner_progress_services.py:
```diff
@@ -972,9 +972,8 @@ def get_activity_progress(user_id):
-        list(str). The titles of the collections to which new explorations have
-            been added.
-
+            The third return valus is the titles of the collections to which new
```
valus --> value

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/CollectionSummaryTileDirective.js:
```diff
@@ -35,7 +35,7 @@ oppia.directive('collectionSummaryTile', [
         isLinkedToEditorPage: '=?isLinkedToEditorPage',
         getCategory: '&category',
-        isPlaylistMode: '&playlistMode',
+        isPlaylistTile: '&playlistTile',
```
I think you want the actual directive attr name to also be is-playlist-tile, otherwise it's as though it's naming the playlist tile that's being passed in.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/components/summary_tile/ExplorationSummaryTileDirective.js:
```diff
@@ -44,7 +44,7 @@ oppia.directive('explorationSummaryTile', [
         // desktop version of the summary tile is always displayed.
         mobileCutoffPx: '@mobileCutoffPx',
-        isPlaylistMode: '&playlistMode',
+        isPlaylistTile: '&playlistTile',
```
Ditto.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerDashboardIconsDirective.js:
```diff
@@ -33,13 +33,15 @@ oppia.directive('learnerDashboardIcons', [
-                 LearnerPlaylistService) {
-          $scope.activityIsActive = true;
+        function(
+          $scope, LearnerDashboardIdsBackendApiService,
```
Indent this and next two lines by 2.

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/domain/learner_dashboard/LearnerDashboardIconsDirective.js:
```diff
@@ -33,13 +33,15 @@ oppia.directive('learnerDashboardIcons', [
 
           $scope.$watch('activityActive', function(value) {
-            $scope.activityIsActive = !$scope.activityIsActive;
+            $scope.activityIsCurrentlyHoveredOver = (
```
Can we find some way to "set", and not toggle?

arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/learner_dashboard/LearnerDashboard.js:
```diff
@@ -412,8 +409,9 @@ oppia.controller('LearnerDashboard', [
-          'subsectionName', function($scope, $modalInstance, $http,
-            sectionNameI18nId, subsectionName) {
+          'subsectionName', function(
+            $scope, $modalInstance, $http, sectionNameI18nId,
```
Indent this and the next line by 2.

arunabh98 replied:
Done!

arunabh98 commented:
Did the CSS changes.

arunabh98 commented:
Addressed the review comments.

arunabh98 committed:
Address review comments 4

seanlip reviewed (approved):
LGTM from code perspective. Thanks @Arunabh98!

arunabh98 committed:
Address review comments 5.

jaredsilver reviewed (approved):
UI (mostly) LGTM. I'll open a couple issues related to the LD project that would be good for folks to address (ideally before the next release), but this is sufficient for merge 👍 Congrats, @Arunabh98!
